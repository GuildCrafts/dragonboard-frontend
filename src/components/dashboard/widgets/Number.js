import React, { Component } from 'react'

import Header from './Header'

import '../../../styles/dashboard/widget/number.css'
import '../../../styles/dashboard/widget/widget.css'

export default class Number extends Component {

  displayStatistic() {
    const { statistic } = this.props

    return statistic % 1 === 0 ? statistic : statistic.toFixed( 2 )
  }

  completedClass() {
    const { statistic, goal } = this.props

    return statistic / goal >= 1 ? ' complete' : ''
  }

  progress() {
    const { statistic, goal } = this.props

    return `${( statistic / goal * 100 ).toFixed( 2 )}%`
  }

  render() {
    const { title, x, y, statistic, goal } = this.props
    const positionStyle = { top: x, left: y }

    return (
      <article>
        <div className="widget widget-width-1 widget-height-1"
          style={ positionStyle }>
          <Header title={ title }/>

          <div className={ `widget-body stat-wrapper${ this.completedClass() }` }>
            <div className="statistic">{ this.displayStatistic() }</div>

            <div className="goal-progress-wrapper">
              <div className="progress">{ this.progress() }</div>
              <div className="goal">{ goal }</div>
            </div>

            <div className="progress-bar">
              <span style={ { width: `${ this.progress() }` } }></span>
            </div>
          </div>
        </div>
      </article>
    )
  }
}
