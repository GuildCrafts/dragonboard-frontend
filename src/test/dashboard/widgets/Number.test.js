import React from 'react'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'

import Number from '../../../components/dashboard/widgets/Number'
import widgetData from
 '../../../reducers/widgets/dummy-data/dashboard-tools/number.js'

describe( 'Number widget component', () => {
  const givenStatistic = widgetData.statistic
  const givenGoal = widgetData.goal
  const expectedProgress = ( givenStatistic / givenGoal ) * 100
  const roundedProgress = expectedProgress.toFixed( 2 )

  const numberWrapper = shallow( <Number { ...widgetData } /> )

  it( 'renders without crashing', () => {
    const numberSnapshot = renderer.create(
      <Number { ...widgetData } />
    ).toJSON()
    expect( numberSnapshot ).toMatchSnapshot()
  })

  it( 'renders the main statistic', () => {
    const statisticDiv = numberWrapper.find( '.statistic' )
    expect( statisticDiv.type() ).toEqual( 'div' )
    expect( statisticDiv.text() ).toEqual( givenStatistic.toString() )
  })

  it( 'renders the goal', () => {
    const goalDiv = numberWrapper.find( '.goal' )
    expect( goalDiv.type() ).toEqual( 'div' )
    expect( goalDiv.text() ).toEqual( givenGoal.toString() )
  })

  it( 'renders the progress', () => {
    const progressDiv = numberWrapper.find( '.progress' )
    expect( progressDiv.type() ).toEqual( 'div' )
    expect( progressDiv.text() ).toEqual( `${ roundedProgress }%` )
  })

  it( 'calculates the progress into a percentage', () => {
    const progress = numberWrapper.instance().progress()
    expect( progress ).toEqual( `${ roundedProgress }%` )
  })

})
