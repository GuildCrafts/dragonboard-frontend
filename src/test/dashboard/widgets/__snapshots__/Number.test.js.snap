exports[`Number widget component renders without crashing 1`] = `
<article>
  <div
    className="widget widget-width-1 widget-height-1"
    style={
      Object {
        "left": "500px",
        "top": "10px",
      }
    }>
    <header>
      <span>
        <label
          className="header-text">
           
          Number
           
        </label>
      </span>
    </header>
    <div
      className="widget-body stat-wrapper">
      <div
        className="statistic">
        151
      </div>
      <div
        className="goal-progress-wrapper">
        <div
          className="progress">
          75.50%
        </div>
        <div
          className="goal">
          200
        </div>
      </div>
      <div
        className="progress-bar">
        <span
          style={
            Object {
              "width": "75.50%",
            }
          } />
      </div>
    </div>
  </div>
</article>
`;
